x/metadata/client/cli/tx.go

```diff
@@ -242,3 +250,163 @@ func ModifyOsLocatorCmd() *cobra.Command {
+		Short: "Add/Update metadata contract specification on the provenance blockchain",
+		Long: `Add/Update metadata contract specification on the provenance blockchain
+[contractspec-id] - metaaddress of contract specification
+[owners] - comma delimited list of bech32 owner addresses
+[parties-involved] - comma delimited list of party types.  Accepted values: originator,servicer,investor,custodian,owner,affiliate,omnibus,provenance
+[source-type] - accepted values: hash or resourceid
+[source-value] - source identifier of type hash or resourceid
+[classname] - Name of contract specification
+[description-name] - optional- description name identifier 
+[description] - optional - description text
+[website-url] - optional - address of website
+[icon-url] - optional - address to a image to be used as an icon
+		`,
+		Args: cobra.RangeArgs(6, 10),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx, err := client.GetClientTxContext(cmd)
+			if err != nil {
+				return err
+			}
+
+			var specificationID types.MetadataAddress
+			specificationID, err = types.MetadataAddressFromBech32(args[0])
+			if err != nil {
+				return err
+			}
+
+			if !specificationID.IsContractSpecificationAddress() {
+				return fmt.Errorf("invalid contract specification id: %s", args[0])
+			}
+
+			partiesInvolved := parsePartyTypes(args[2])
+			description := parseDescription(args[6:])
+			contractSpecification := types.ContractSpecification{SpecificationId: specificationID,
+				Description:     description,
+				OwnerAddresses:  strings.Split(args[1], ","),
+				PartiesInvolved: partiesInvolved,
+				ClassName:       args[5],
+			}
+			switch s := strings.ToUpper(args[3]); s {
+			case "RESOURCEID":
+				var recordID types.MetadataAddress
+				recordID, err = types.MetadataAddressFromBech32(args[4])
+				if err != nil {
+					return err
+				}
+				contractSpecification.Source = &types.ContractSpecification_ResourceId{
+					ResourceId: recordID,
+				}
+			case "HASH":
+				contractSpecification.Source = &types.ContractSpecification_Hash{
+					Hash: args[4],
+				}
+			default:
+				return fmt.Errorf("incorrect source type for contract specification: %s", s)
+			}
+
+			signers := parseSigners(cmd, &clientCtx)
+
+			msg := types.NewMsgAddContractSpecificationRequest(contractSpecification, signers)
+			err = msg.ValidateBasic()
+			if err != nil {
+				return err
+			}
+
+			return tx.GenerateOrBroadcastTxCLI(clientCtx, cmd.Flags(), msg)
+		},
+	}
+
+	addSignerFlagCmd(cmd)
+	flags.AddTxFlagsToCmd(cmd)
+
+	return cmd
+}
+
+func addSignerFlagCmd(cmd *cobra.Command) {
+	cmd.Flags().String(FlagSigners, "", "comma delimited list of bech32 addresses")
+}
+
+func parseSigners(cmd *cobra.Command, client *client.Context) []string {
+	flagSet := cmd.Flags()
+	if flagSet.Changed(FlagSigners) {
+		signerList, _ := flagSet.GetString(FlagSigners)
+		return strings.Split(signerList, ",")
+	}
+	return []string{client.GetFromAddress().String()}
```

Comment:
This could alternatively remove the from address from the signers (if present) and insert the from address as the first address in the list... that would support use of both flags together.  If thats not done then you would need an exclusive switch to prevent an invalid operation where it signs with potentially the wrong key